contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -0,0 +1,832 @@
+		points.GetNormalsAttr().ValueMightBeTimeVarying() ||
+		points.GetVelocitiesAttr().ValueMightBeTimeVarying() ||
+		points.GetWidthsAttr().ValueMightBeTimeVarying() ||
+		points.GetIdsAttr().ValueMightBeTimeVarying() ||
+		hasTimeVaryingPrimVars( points );
+}
+
+bool isTimeVarying( pxr::UsdPrim prim )
+{
+
+	if( pxr::UsdGeomMesh mesh = pxr::UsdGeomMesh::Get( prim.GetStage(), prim.GetPath() ) )
+	{
+		return isTimeVarying( mesh );
+	}
+
+	if( pxr::UsdGeomPoints points = pxr::UsdGeomPoints::Get( prim.GetStage(), prim.GetPath() ) )
+	{
+		return isTimeVarying( points );
+	}
+
+	if( pxr::UsdGeomCurves curves = pxr::UsdGeomCurves::Get( prim.GetStage(), prim.GetPath() ) )
+	{
+		return isTimeVarying( curves );
+	}
+
+	return false;
+}
+
+} // namespace
+
+class USDScene::IO : public RefCounted
+{
+	public:
+		IO( const std::string &fileName ) : m_fileName( fileName )
+		{
+		}
+
+		const std::string &fileName() const
+		{
+			return m_fileName;
+		}
+
+		virtual pxr::UsdPrim root() const = 0;
+		virtual pxr::UsdTimeCode getTime( double timeSeconds ) const = 0;
+	private:
+		std::string m_fileName;
+};
+
+class USDScene::Reader : public USDScene::IO
+{
+	public:
+		Reader( const std::string &fileName ) : IO( fileName )
+		{
+			m_usdStage = pxr::UsdStage::Open( fileName );
+			m_timeCodesPerSecond = m_usdStage->GetTimeCodesPerSecond();
+			m_rootPrim = m_usdStage->GetPseudoRoot();
+		}
+
+		pxr::UsdPrim root() const override
+		{
+			return m_rootPrim;
+		}
+
+		pxr::UsdTimeCode getTime( double timeSeconds ) const override
+		{
+			return timeSeconds * m_timeCodesPerSecond;
+		}
+
+	private:
+		pxr::UsdStageRefPtr m_usdStage;
+		pxr::UsdPrim m_rootPrim;
+
+		double m_timeCodesPerSecond;
+};
+
+class USDScene::Writer : public USDScene::IO
+{
+	public:
+		Writer( const std::string &fileName ) : IO( fileName )
+		{
+			m_usdStage = pxr::UsdStage::CreateNew( "gaffer" );
+			m_rootPrim = m_usdStage->GetPseudoRoot();
+		}
+
+		pxr::UsdPrim root() const override
+		{
+			return m_rootPrim;
+		}
+
+		pxr::UsdTimeCode getTime( double timeSeconds ) const override
+		{
+			return 0.0;
+		}
+
+	private:
+		pxr::UsdStageRefPtr m_usdStage;
+		pxr::UsdPrim m_rootPrim;
+};
+
+USDScene::USDScene( const std::string &path, IndexedIO::OpenMode &mode )
+{
+	switch( mode )
+	{
+		case IndexedIO::Read :
+			m_root = new Reader( path );
+			m_prim = m_root->root();
+			break;
+		case IndexedIO::Write :
+			m_root = new Writer( path );
+			m_prim = m_root->root();
+			break;
+		default:
+			throw Exception( " Unsupported OpenMode " );
+	}
+}
+
+USDScene::USDScene( IOPtr root, pxr::UsdPrim prim ) : m_root( root ), m_prim( prim )
+{
+
+}
+
+USDScene::~USDScene()
+{
+}
+
+std::string USDScene::fileName() const
+{
+	return m_root->fileName();
+}
+
+Imath::Box3d USDScene::readBound( double time ) const
+{
+	pxr::TfTokenVector purposes;
+	purposes.push_back( pxr::UsdGeomTokens->default_ );
+	pxr::UsdGeomBBoxCache bboxCache( m_root->getTime( time ), purposes );
```

Comment:
A cache _sounds_ like it should be persisted, rather than recreated each time we want to query something new. Is that not the case?